astropy_healpix/_core.c

```diff
@@ -0,0 +1,368 @@
+{
+    npy_intp i, n = dimensions[0];
+
+    for (i = 0; i < n; i ++)
+    {
+        double  lon    = *(double *)  &args[0][i * steps[0]];
+        double  lat    = *(double *)  &args[1][i * steps[1]];
+        int     nside  = *(int *)     &args[2][i * steps[2]];
+        int64_t indices[4];
+        double weights[4];
+        int j;
+
+        interpolate_weights(lon, lat, indices, weights, nside);
+        for (j = 0; j < 4; j ++)
+        {
+            *(int64_t *) &args[3 + j][i * steps[3 + j]] = indices[j];
+            *(double *)  &args[7 + j][i * steps[7 + j]] = weights[j];
+        }
+    }
+}
+
+
+static void neighbours_loop(
+    char **args, npy_intp *dimensions, npy_intp *steps, void *data)
+{
+    order_funcs *funcs = data;
+    npy_intp i, n = dimensions[0];
+
+    for (i = 0; i < n; i ++)
+    {
+        int64_t pixel = *(int64_t *) &args[0][i * steps[0]];
+        int     nside = *(int *)     &args[1][i * steps[1]];
+        int64_t neighbours[] = {-1, -1, -1, -1, -1, -1, -1, -1};
+        int j;
+        int64_t xy = -1;
+
+        if (pixel_nside_valid(pixel, nside))
+            xy = funcs->order_to_xy(pixel, nside);
+        if (xy >= 0)
+            healpixl_get_neighbours(xy, neighbours, nside);
+
+        for (j = 0; j < 8; j ++)
+        {
+            int k = 4 - j;
+            if (k < 0)
+                k += 8;
+            xy = neighbours[k];
+            if (xy >= 0)
+                pixel = funcs->xy_to_order(xy, nside);
+            else {
+                pixel = -1;
+                _npy_set_floatstatus_invalid();
+            }
+            *(int64_t *) &args[2 + j][i * steps[2 + j]] = pixel;
+        }
+    }
+}
+
+
+static PyObject *healpix_cone_search(
+    PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    PyObject *result = NULL;
+    static const char *kws[] = {"lon", "lat", "radius", "nside", "order", NULL};
+    double lon, lat, radius;
+    int nside;
+    char *order;
+    int64_t *indices, n_indices;
+    int64_t *result_data;
+
+    if (!PyArg_ParseTupleAndKeywords(
+            args, kwargs, "dddis", kws, &lon, &lat, &radius, &nside, &order))
+        return NULL;
+
+    n_indices = healpix_rangesearch_radec_simple(
+        lon, lat, radius, nside, 0, &indices);
+    if (!indices)
+    {
+        PyErr_SetString(
+            PyExc_RuntimeError, "healpix_rangesearch_radec_simple failed");
+        goto done;
+    }
+
+    {
+        npy_intp dims[] = {n_indices};
+        result = PyArray_SimpleNew(1, dims, NPY_INT64);
+    }
+    if (!result)
+        goto done;
+
+    result_data = PyArray_DATA((PyArrayObject *) result);
+
+    if (strcmp(order, "nested") == 0)
+    {
+        int i;
+        for (i = 0; i < n_indices; i ++)
+            result_data[i] = healpixl_xy_to_nested(indices[i], nside);
+    } else {
+        int i;
+        for (i = 0; i < n_indices; i ++)
+            result_data[i] = healpixl_xy_to_ring(indices[i], nside);
+    }
+
+done:
```

Comment:
I find `healpix_cone_search` a bit complex. Maybe replace the two `goto` with simple `return`?
I think in the first case `free(indices)` isn't needed because it's in an `if (!indices)`. So that would really just leave one use for the goto, and you might as well just use normal return?

Reply:
Fixed.